onert-micro/onert-micro/src/core/train/OMTrainingHandler.cpp

```diff
@@ -0,0 +1,277 @@
+    if (forward_output_tensor->type() != circle::TensorType_FLOAT32)
+      return UnsupportedType;
+
+    // Get calculated data
+    uint8_t *calculated_data = nullptr;
+    OMStatus status = forward_storage.getDataByTensorIndex(&calculated_data, forward_output_index);
+    if (status != Ok)
+      return status;
+    assert(calculated_data != nullptr);
+
+    // Get target data
+    auto data_type_size = sizeof(core::OMDataType(forward_output_tensor->type()));
+    size_t offset = batch_num * data_type_size * flat_size;
+    uint8_t *target_data = _training_storage.getTargetData(i) + offset;
+    OMLoss loss_type = config.training_context.loss;
+
+    // Allocate data for error gradient for current calculated data and target data
+    uint8_t *output_grad_data;
+    core::memory::OMMemoryManager::allocateMemory(flat_size * data_type_size, &output_grad_data);
+    // Save error gradient into backward storage associated with current output tensor index
+    backward_storage.saveDataToTensorIndex(output_grad_data, forward_output_index);
+
+    // Handle different loss types
+    switch (loss_type)
+    {
+      case MSE:
+      {
+        losses_functions::MSE::calculateErrorBackpropagation(
+          flat_size, reinterpret_cast<float *>(calculated_data),
+          reinterpret_cast<float *>(target_data), reinterpret_cast<float *>(output_grad_data));
+        break;
+      }
+      case CROSS_ENTROPY:
+      {
+        losses_functions::CrossEntropy::calculateErrorBackpropagation(
+          flat_size, reinterpret_cast<float *>(calculated_data),
+          reinterpret_cast<float *>(target_data), reinterpret_cast<float *>(output_grad_data));
+        break;
+      }
+      default:
+      {
+        assert(false && "Unsupported loss type");
+        return UnsupportedType;
+      }
+    }
+  }
+
+  return Ok;
+}
+
+/*
+ * Update weights with current optimizer logic
+ */
+OMStatus OMTrainingHandler::updateWeights(const OMConfig &config, OMRuntimeContext &context)
+{
+  OMStatus status = Ok;
+
+  // Chose optimizer type
+  switch (config.training_context.optimizer)
+  {
+    case SGD:
+    {
+      auto *sgd_optimizer = _training_storage.getSGD();
+      assert(sgd_optimizer != nullptr);
+      if (sgd_optimizer == nullptr)
+        return UnknownError;
+
+      status = sgd_optimizer->updateWeights(config.training_context, context);
+      assert(status == Ok);
+      // Reset values
+      sgd_optimizer->reset();
+      break;
+    }
+    case ADAM:
+    {
+      auto *adam_optimizer = _training_storage.getAdam();
+      assert(adam_optimizer != nullptr);
+      if (adam_optimizer == nullptr)
+        return UnknownError;
+
+      status = adam_optimizer->updateWeights(config.training_context, context);
+      assert(status == Ok);
+      // Reset values
+      adam_optimizer->reset();
+      break;
+    }
+    default:
+    {
+      assert(false && "Unsupported type");
+      return UnsupportedType;
+    }
+  }
+
+  return status;
+}
```

Comment:
If we make a map for 
{ config.training_context.optimizer : std::unique_ptr<opimizer>... }
I think we can remove all these switch cases and code dependencies like `.getSGD`, `getAdam`... for later extensibility.

Does it have a risk? or costs?

Reply:
Yes, I think it is good idea, thank you. I will try it now and push this changes :)

Reply:
@ljwoo94 
Ah to save it in `std::unordered_map<optimizer_type , std::unique_ptr<Optimizer>>` I need common `Optimizer` type and virtual mechanism to call right functions. But in onert-micro we disable virtual mechanism in order to save binary. 
So, I think such switch case is the most readable way to use different optimizers :(
 Or do you have some ideas?

Reply:
> Or do you have some ideas?
I guess not :'(

Approved PR. LGTM :)